Warn about missing OAuth scopes when reporting HTTP 4xx errors

mislav opened the pull request:
If a 4xx server response lists scopes in the `X-Accepted-Oauth-Scopes` header that are not present in the `X-Oauth-Scopes` header, the final error messaging on stderr will now include a hint for the user that they might need to request the additional scope:

    $ gh codespace list
    error getting codespaces: HTTP 403: Must have admin rights to Repository. (https://api.github.com/user/codespaces?per_page=30)
    This API operation needs the "codespace" scope. To request it, run:  gh auth refresh -h github.com -s codespace

Fixes #4464, fixes #4518
Ref. https://github.com/cli/cli/pull/4451 https://github.com/cli/cli/issues/4415

Note that not all API endpoints consistently report the scopes they need. For example, the gist create endpoint doesn't report that it needs the `gist` scope, nor does the repo delete endpoint report that it needs `delete_repo`.

mislav committed:
Warn about missing OAuth scopes when reporting HTTP 4xx errors

If a 4xx server response lists scopes in the X-Accepted-Oauth-Scopes
header that are not present in the X-Oauth-Scopes header, the final
error messaging on stderr will now include a hint for the user that they
might need to request the additional scope:

    $ gh codespace list
    error getting codespaces: HTTP 403: Must have admin rights to Repository. (https://api.github.com/user/codespaces?per_page=30)
    This API operation needs the "codespace" scope. To request it, run:  gh auth refresh -h github.com -s codespace

mislav committed:
Ensure NOT_FOUND error when querying private repos using insufficient scope

mislav committed:
Consistent error handling in codespaces API operations

Using HandleHTTPError ensures that hints regarding insufficient OAuth
scopes will be properly reported on stderr.

mislav committed:
Remove OAuth scopes checking logic from `ssh-key` commands

Scopes checking is now handled on the HTTP client level for all commands.

samcoe commented on api/client_test.go:
```diff
@@ -146,3 +146,67 @@ func TestHandleHTTPError_GraphQL502(t *testing.T) {
+			resp: makeResponse(500, "https://api.github.com/gists", "repo", "gist"),
+			want: ``,
+		},
+		{
+			name: "no scopes on token",
+			resp: makeResponse(404, "https://api.github.com/gists", "", "gist, delete_repo"),
+			want: ``,
+		},
```
I don't exactly understand this case, wouldn't we want to report that the token is missing the two required scopes here?

mislav replied:
A token that the server reports as having no scopes whatsoever is most likely an Apps token (not an OAuth token or a PAT). In that case we give up on performing any scopes checking.

samcoe commented on api/queries_repo.go:
```diff
@@ -241,12 +241,23 @@ func FetchRepository(client *Client, repo ghrepo.Interface, fields []string) (*R
 		return nil, err
 	}
-	return InitRepoHostname(&result.Repository, repo.RepoHost()), nil
+	// The GraphQL API should have returned an error in case of a missing repository, but this isn't
+	// guaranteed to happen when an authentication token with insufficient permissions is being used.
+	if result.Repository == nil {
+		return nil, GraphQLErrorResponse{
+			Errors: []GraphQLError{{
+				Type:    "NOT_FOUND",
+				Message: fmt.Sprintf("Could not resolve to a Repository with the name '%s/%s'.", repo.RepoOwner(), repo.RepoName()),
+			}},
+		}
+	}
```
Is this done on purpose in the GraphQL endpoint for security reasons? If so, are we circumventing that and exposing that a repository does not exist? Would it be better to return a response that the user has insufficient permissions?

mislav replied:
This is compensating for what is likely a bug in the GraphQL API. If a repository couldn't be found, GraphQL was supposed to generate an error, but sometimes it does not. This makes sure that an error is always raised in the case of a missing repository.

Right now we don't check for insufficient permissions in this endpoint, but we could! For example, if the token only has `public_repo` scope and a repository couldn't be found, we could point out that you are unable to fetch private repos until you enable the `repo` scope. Do you think it would be worth adding that check on top of what's already here?

samcoe replied:
Got it. I think it would be a nice improvement for the future, but not necessary to land this PR.

samcoe commented on api/queries_repo.go:
```diff
@@ -280,16 +291,24 @@ func GitHubRepo(client *Client, repo ghrepo.Interface) (*Repository, error) {
+	if err := client.GraphQL(repo.RepoHost(), query, variables, &result); err != nil {
 		return nil, err
 	}
+	// The GraphQL API should have returned an error in case of a missing repository, but this isn't
+	// guaranteed to happen when an authentication token with insufficient permissions is being used.
+	if result.Repository == nil {
+		return nil, GraphQLErrorResponse{
+			Errors: []GraphQLError{{
+				Type:    "NOT_FOUND",
+				Message: fmt.Sprintf("Could not resolve to a Repository with the name '%s/%s'.", repo.RepoOwner(), repo.RepoName()),
+			}},
+		}
+	}
```
Same questions as above.

samcoe reviewed (approved):
Code looks good. I don't have any real context on the CS file changes but they looked like straight forward replacements. I left some questions for discussion, but don't think they are blockers.

mislav committed:
auth refresh: preserve existing scopes when requesting new ones

When there was a previously valid token that was granted some scopes,
ensure all those scopes will be re-requested when doing the
authentication flow for the new token.
